### PR TITLE
Recover(), Catcher, and WaitGroup

### PR DESCRIPTION
**Pull request by amery**

### CODE REVIEW

**Comment by amery** [comment]
WaitGroup uses Catcher, Catcher uses Recover

**Comment by amery** [comment]
after this, the only other thing I would like to add is Frame/Stack, and add a Stack to panicError.

**Comment by amery** [comment]
PanicError interface renamed to Recovered

File: panic.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
+package core
+
+import (
+	"fmt"
+	"sync/atomic"
+)
+
+// Recovered is a caught error from a panic call
```
Comment (karasz) [inline panic.go:8]: // Recover is an error caught from a panic call...

File: panic.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+		return err
+	}
+
+	// all good
+	return nil
+}
+
+// Try calls a function, returning its organic error,
+// or storing the recovered error for later consumption
+func (p *Catcher) Try(fn func() error) error {
```
Comment (karasz) [inline panic.go:73]: of God, oh God, Java looks back at me....

Comment (amery) [reply panic.go:73]: I could make it private or change the name @karasz

**Comment by amery** [comment]
@karasz English bug fixed

File: panic.go (hunk 2 above)
Comment (karasz) [reply panic.go:73]: Nah. its fine.

File: sync.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	onError func(error)
+}
+
+// OnError sets a helper that will be called when
+// the first worker returns an error or panics
+func (wg *WaitGroup) OnError(fn func(error)) {
+	wg.onError = fn
+}
+
+// Go spawns a supervised gorouting
```
Comment (karasz) [inline sync.go:21]: s/gorouting/goroutine/

**Review by karasz — CHANGES_REQUESTED** [review]
some typos need to be solved

File: sync.go (hunk 3 above)
Comment (amery) [reply sync.go]: thanks

**Comment by amery** [comment]
typos corrected. PR amended

**Review by karasz — APPROVED** [review]
LGTM

**Review by karasz — APPROVED** [review]
LGTM

**Comment by amery** [comment]
`v0.3.0` tagged

**Pull request merged**